DQMServices/Core/interface/locking_ptr.h

```diff
@@ -0,0 +1,136 @@
+#ifndef locking_ptr_h
+#define locking_ptr_h
+
+#include <memory>
+
+namespace detail {
+
+  class lock_base
+  {
+  public:
+    lock_base() = default;
+    virtual ~lock_base() = default;
+
+    virtual void lock() const = 0;
+    virtual void unlock() const = 0;
+  };
+
+  template <typename L>
+  class lock_impl : public lock_base
```

Comment:
Isn't it true that only the `MonitorElement` and `locking_ptr` have to agree on the underlying lock type? If so, why not just declare the type as a `using` in `locking_ptr` and have `MonitorElement` explicitly use that type? That should solve the maintenance case and make the implementation of `locking_ptr` simpler. I think it would also avoid the need to use the heap each time a `locking_ptr` is made.